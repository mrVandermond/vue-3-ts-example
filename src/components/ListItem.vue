<template>
  <div
    :class="{
      internal: props.isInternal,
      'is-open': isOpen
    }"
    class="list-item"
  >
    <div
      class="list-item-header"
      @click="onClickHeader"
    >
      <ListItemButton
        :is-open="isOpen"
        class="list-item-button"
      />

      <slot name="label" />
    </div>

    <div
      :class="{
        'is-loading': props.isLoadingContent,
      }"
      class="list-item-content"
    >
      <Loader v-if="props.isLoadingContent" />

      <Error
        v-else-if="!props.isLoadingContent && props.isFailLoadingContent"
        :is-title-from-right="true"
      />

      <slot
        v-else
        name="content"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue';
import ListItemButton from '@/components/ListItemButton.vue';
import Loader from '@/components/Loader.vue';
import Error from '@/components/Error.vue';

const isOpen = ref(false);

const props = withDefaults(defineProps<{
  isInternal?: boolean;
  isLoadingContent?: boolean;
  isFailLoadingContent?: boolean;
}>(), {
  isInternal: false,
  isLoadingContent: false,
  isFailLoadingContent: false,
});
const emit = defineEmits<{
  (event: 'click', isOpen: boolean): void;
}>();

const onClickHeader = (): void => {
  isOpen.value = !isOpen.value;
  emit('click', isOpen.value);
};
</script>

<style lang="scss">
.list-item {
  padding: 24px 0;

  &.internal {
    margin-left: 56px;
  }

  &.is-open {
    padding-bottom: 0;
  }
}
.list-item-button {
  margin-right: 24px;
}
.list-item-header {
  cursor: pointer;
  display: flex;
  align-items: center;
}
.list-item-content.is-loading {
  min-height: 320px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
}
</style>
