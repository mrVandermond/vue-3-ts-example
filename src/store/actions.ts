import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
