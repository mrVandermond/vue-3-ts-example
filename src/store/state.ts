import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';

export interface IState {
  favoritePhotos: IPhotoWithIsFavorite[];
  modal: IModalConfig;
  title: ITitleConfig;
}

export const state = (): IState => ({
  favoritePhotos: [],
  modal: {
    isOpen: false,
    url: '',
  },
  title: {
    isShow: false,
    text: '',
    xCoord: 0,
    yCoord: 0,
  },
});
