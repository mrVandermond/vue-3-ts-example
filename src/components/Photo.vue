<template>
  <div
    class="photo"
    :data-title="props.photo.title"
    @click="onClickOpenModal"
  >
    <PhotoStar
      :is-active="props.photo.isFavorite"
      class="photo-star"
      @click.stop="onClickStar"
    />

    <img
      :src="props.photo.thumbnailUrl"
      alt="photo-image"
    >
  </div>
</template>

<script lang="ts" setup>
import PhotoStar from '@/components/PhotoStar.vue';
import type { IPhotoWithIsFavorite } from '@/types';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const props = defineProps<{
  photo: IPhotoWithIsFavorite;
}>();
const emit = defineEmits<{
  (event: 'addedToFavorites', photo: IPhotoWithIsFavorite): void;
  (event: 'removeFromFavorites', photo: IPhotoWithIsFavorite): void;
}>();
const store = customUseStore();

const onClickStar = (): void => {
  if (props.photo.isFavorite) {
    emit('removeFromFavorites', props.photo);

    return;
  }

  emit('addedToFavorites', props.photo);
};
const onClickOpenModal = (): void => {
  store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
    isOpen: true,
    url: props.photo.url,
  });
};
</script>

<style lang="scss">
.photo {
  overflow: hidden;
  border-radius: 4px;
  width: 150px;
  height: 150px;
  cursor: pointer;
  position: relative;
}
.photo-star {
  position: absolute;
  top: 8px;
  right: 8px;
}
</style>
