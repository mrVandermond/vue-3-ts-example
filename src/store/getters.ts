import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
