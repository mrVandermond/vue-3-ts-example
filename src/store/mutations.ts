import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
