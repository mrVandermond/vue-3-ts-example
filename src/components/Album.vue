<template>
  <ListItem
    :is-loading-content="isLoadingPhotos"
    :is-fail-loading-content="isFailLoadingPhotos"
    is-internal
    @click="onClickItem"
  >
    <template #label>
      <div class="album-label">
        {{ props.album.title }}
      </div>
    </template>

    <template
      v-if="isOpenAlbum"
      #content
    >
      <div
        class="list-photo"
        @mousemove.passive="titleOptions.onMouseMove"
      >
        <Photo
          v-for="photo in photos"
          :key="photo.id"
          :photo="photo"
          @added-to-favorites="onAddedToFavorites"
          @remove-from-favorites="onRemoveFromFavorites"
        />
      </div>
    </template>
  </ListItem>
</template>

<script lang="ts" setup>
import type { IAlbum, IPhotoWithIsFavorite } from '@/types';
import ListItem from '@/components/ListItem.vue';
import Photo from '@/components/Photo.vue';
import { ref, watchEffect } from 'vue';
import { storeTypes } from '@/store/types';
import { customUseStore } from '@/store';
import { useTitle } from '@/composables';

const props = defineProps<{
  album: IAlbum;
}>();

const photos = ref<IPhotoWithIsFavorite[]>([]);
const isLoadingPhotos = ref(false);
const isFailLoadingPhotos = ref(false);
const isOpenAlbum = ref(false);

const store = customUseStore();
const titleOptions = useTitle();

watchEffect(() => {
  store.commit(storeTypes.EMutations.SET_TITLE_CONFIG, {
    isShow: titleOptions.isShowTitle.value,
    xCoord: titleOptions.xCoordMouse.value,
    yCoord: titleOptions.yCoordMouse.value,
    text: titleOptions.titleText.value,
  });
});

function toggleIsFavorite(index: number): void {
  photos.value[index] = {
    ...photos.value[index],
    isFavorite: !photos.value[index].isFavorite,
  };
}

async function onClickItem(isOpen: boolean): Promise<void> {
  isOpenAlbum.value = isOpen;

  if (!isOpen) return;

  isLoadingPhotos.value = true;

  try {
    const photosWithoutIsFavorite = await store.dispatch(storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID, props.album.id);

    photos.value = photosWithoutIsFavorite.map((photo) => ({
      ...photo,
      isFavorite: store.getters[storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](photo.id),
    }));
  } catch (error) {
    isFailLoadingPhotos.value = true;
  }

  isLoadingPhotos.value = false;
}
function onAddedToFavorites(photo: IPhotoWithIsFavorite): void {
  store.commit(storeTypes.EMutations.ADD_TO_FAVORITES, [photo]);

  const photoIndex = photos.value.findIndex((item) => item.id === photo.id);

  if (photoIndex === -1) return;

  toggleIsFavorite(photoIndex);
}
function onRemoveFromFavorites(photo: IPhotoWithIsFavorite): void {
  store.commit(storeTypes.EMutations.REMOVE_FROM_FAVORITES, photo);

  const photoIndex = photos.value.findIndex((item) => item.id === photo.id);

  if (photoIndex === -1) return;

  toggleIsFavorite(photoIndex);
}
</script>

<style lang="scss">
.album-label {
  font-size: 18px;
}
.list-photo {
  display: flex;
  flex-wrap: wrap;
  gap: 42px;
  padding: 32px 105px 32px 49px;
}
</style>
