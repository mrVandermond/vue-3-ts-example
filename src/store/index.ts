import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
