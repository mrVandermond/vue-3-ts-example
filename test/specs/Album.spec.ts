import Album from '@/components/Album.vue';
import ListItem from '@/components/ListItem.vue';
import Photo from '@/components/Photo.vue';

import type { Store } from '@/store/types/overrides';
import type { IAlbum, IPhoto, IUser } from '@/types';
import type { IUseTitleOptions } from '@/composables/title';
import type { SpyInstance } from 'vitest';

import { nextTick, ref } from 'vue';
import { flushPromises } from '@vue/test-utils';
import { customCreateStore, key } from '@/store';
import { storeTypes } from '@/store/types';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';

import { vi } from 'vitest';
import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

const album: IAlbum = {
  id: 1,
  title: 'test',
  userId: 2,
};
const examplePhotos: IPhoto[] = [
  {
    id: 1,
    title: 'photo',
    url: 'https://url',
    albumId: 1,
    thumbnailUrl: 'https://thumbnailUrl',
  },
];
const hoistedVars = vi.hoisted(() => ({
  onMouseMoveMock: vi.fn(),
}));

vi.mock('../../src/composables/index', () => ({
  useTitle: vi.fn((): IUseTitleOptions => ({
    onMouseMove: hoistedVars.onMouseMoveMock,
    isShowTitle: ref(false),
    xCoordMouse: ref(0),
    yCoordMouse: ref(0),
    titleText: ref(''),
  })),
}));

describe('Album component', () => {
  let store!: Store;
  const getWrapper = (): TWrapperFactoryReturnType => wrapperFactory(Album, {
    props: {
      album,
    },
    global: {
      plugins: [[store, key]],
      stubs: {
        ListItem,
      },
    },
  });

  let commitSpy: SpyInstance;
  let dispatchSpy: SpyInstance;

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });
    commitSpy = vi.spyOn(store, 'commit');
    dispatchSpy = vi.spyOn(store, 'dispatch');
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    vi.unmock('@/composables/title.ts');
  });

  it('Отображение в закрытом состоянии', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('Отображение в открытом состоянии', () => {
    beforeEach(() => {
      dispatchSpy.mockResolvedValue(Promise.resolve(examplePhotos) as unknown as Promise<IUser[]>);
    });

    const openListItem = async (wrapper: TWrapperFactoryReturnType): Promise<void> => {
      const listItem = wrapper.findComponent(ListItem);

      listItem.vm.$emit('click', true);

      await flushPromises();
    };

    it('Фотографии отображаются при клике на ListItem', async () => {
      const wrapper = getWrapper();

      await openListItem(wrapper);

      expect(store.dispatch).toHaveBeenCalledWith(storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID, expect.any(Number));
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.findAllComponents(Photo).length).toBe(1);
    });

    it('При движении мыши вызывается commit', async () => {
      const wrapper = getWrapper();

      await openListItem(wrapper);

      const photo = wrapper.findComponent(Photo);

      await photo.trigger('mousemove');

      expect(hoistedVars.onMouseMoveMock).toHaveBeenCalled();
    });

    it('Проверка обработки события addedToFavorites', async () => {
      const wrapper = getWrapper();

      await openListItem(wrapper);

      const photo = wrapper.findComponent(Photo);

      photo.vm.$emit('addedToFavorites', examplePhotos[0]);

      await nextTick();

      expect(photo.props()).toMatchSnapshot();
      expect(commitSpy.mock.calls[1]).toMatchSnapshot();
    });

    it('Проверка обработки события removeFromFavorites', async () => {
      store.commit(storeTypes.EMutations.ADD_TO_FAVORITES, [{
        ...examplePhotos[0],
        isFavorite: true,
      }]);

      const wrapper = getWrapper();

      await openListItem(wrapper);

      const photo = wrapper.findComponent(Photo);

      photo.vm.$emit('removeFromFavorites', examplePhotos[0]);

      await nextTick();

      expect(photo.props()).toMatchSnapshot();
      expect(commitSpy.mock.calls[2]).toMatchSnapshot();
    });
  });

  it('Отображение при закрытии', async () => {
    dispatchSpy.mockResolvedValue(Promise.resolve(examplePhotos) as unknown as Promise<IUser[]>);

    const wrapper = getWrapper();
    const listItem = wrapper.findComponent(ListItem);

    listItem.vm.$emit('click', true);

    await flushPromises();

    listItem.vm.$emit('click', false);

    await flushPromises();

    expect(dispatchSpy).toHaveBeenCalledTimes(1);
    expect(wrapper.findAllComponents(Photo).length).toBe(0);
  });
});
