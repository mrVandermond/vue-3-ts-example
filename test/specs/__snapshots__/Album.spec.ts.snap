// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Album component > Отображение в закрытом состоянии 1`] = `
"<div class=\\"internal list-item\\">
  <div class=\\"list-item-header\\">
    <list-item-button-stub isopen=\\"false\\" class=\\"list-item-button\\"></list-item-button-stub>
    <div class=\\"album-label\\">test</div>
  </div>
  <div class=\\"list-item-content\\"></div>
</div>"
`;

exports[`Album component > Отображение в открытом состоянии > Проверка обработки события addedToFavorites 1`] = `
{
  "photo": {
    "albumId": 1,
    "id": 1,
    "isFavorite": true,
    "thumbnailUrl": "https://thumbnailUrl",
    "title": "photo",
    "url": "https://url",
  },
}
`;

exports[`Album component > Отображение в открытом состоянии > Проверка обработки события addedToFavorites 2`] = `
[
  "ADD_TO_FAVORITES",
  [
    {
      "albumId": 1,
      "id": 1,
      "thumbnailUrl": "https://thumbnailUrl",
      "title": "photo",
      "url": "https://url",
    },
  ],
]
`;

exports[`Album component > Отображение в открытом состоянии > Проверка обработки события removeFromFavorites 1`] = `
{
  "photo": {
    "albumId": 1,
    "id": 1,
    "isFavorite": false,
    "thumbnailUrl": "https://thumbnailUrl",
    "title": "photo",
    "url": "https://url",
  },
}
`;

exports[`Album component > Отображение в открытом состоянии > Проверка обработки события removeFromFavorites 2`] = `
[
  "REMOVE_FROM_FAVORITES",
  {
    "albumId": 1,
    "id": 1,
    "thumbnailUrl": "https://thumbnailUrl",
    "title": "photo",
    "url": "https://url",
  },
]
`;

exports[`Album component > Отображение в открытом состоянии > Фотографии отображаются при клике на ListItem 1`] = `
"<div class=\\"internal list-item\\">
  <div class=\\"list-item-header\\">
    <list-item-button-stub isopen=\\"false\\" class=\\"list-item-button\\"></list-item-button-stub>
    <div class=\\"album-label\\">test</div>
  </div>
  <div class=\\"list-item-content\\">
    <div class=\\"list-photo\\">
      <photo-stub photo=\\"[object Object]\\"></photo-stub>
    </div>
  </div>
</div>"
`;
